src/puppetlabs/puppetdb/cli/services.clj

```diff
@@ -312,6 +312,37 @@
     (future-cancel command-loader))
   context)
 
+;; A map of required postgres settings and their expected value.
+;; PuppetDB will refuse to start unless ALL of these values
+;; match the actual database configuration settings.
+(def required-pg-settings
+  {:standard_conforming_strings "on"})
+
+(defn verify-database-setting
+  "Given m, a map of database settings, verify 'expected', which has the form
+  [setting-name expected-value]."
+  [m expected]
+  (let [setting (first expected)
+        value (second expected)
+        map-path [setting :setting]]
+    (when-not (= (get-in m map-path) value)
+      {setting {:expected value :actual (get-in m map-path)}})))
+
+(defn verify-database-settings
+  "Ensure the database configuration does not have any settings known
+  to break PuppetDB. If an invalid database configuration is found, throws
+  {:type ::invalid-database-configuration :failed-validation failed-map}"
+  [db-conn-pool]
+  (let [database-settings (jdbc/with-db-connection db-conn-pool (jdbc/query "show all;" ))
+        munged-settings (apply merge
+                                (map (fn [h]
+                                       {(keyword (:name h)) (dissoc h :name)})
+                                     database-settings))
+        invalid-settings (doall (map #(verify-database-setting munged-settings %) required-pg-settings))]
+    (if (seq (filter some? invalid-settings))
+             (throw+ {:type ::invalid-database-configuration :failed-validation
+                      (apply merge invalid-settings)}))))
```

Comment:
Hmm -- might be reasonable to have infrastructure like this from the start, but I offhand, I might have been tempted to just begin with something simpler for this case, i.e.

```
(defn verify-database-settings
  ...
  (let [{:keys[standard_conforming_strings]} settings]
    (when-not (= "on" standard_conforming_strings)
      (throw
        (ex-info "PuppetDB requires the PostgreSQL standard_conforming_strings setting to be \"on\""
                 {:kind ::invalid-database-configuration
                  ...))))
```
Though I suppose that wouldn't make it easy to report multiple errors in one pass.  That could be fixed by just throwing one exception at the end (causing the shutdown), after logging/trs the messages, but it's still not as flexible as what you have.  Not sure -- need to think about it.

Reply:
In any case, I'd argue for avoiding any new slingshot code (since I'd really like to remove it altogether "soon"), i.e. perhaps just be explicit, i.e. throw ex-info like that ^ and then catch ExceptionInfo and use ex-data below.

Reply:
And I'd favor `:kind` rather than `:type` to match our existing precedents.  (Though maybe we use `:type` elsewhere too?)

Reply:
Oh, and I wondered if the setting values might be case sensitive and/or already normalized, i.e. whether "on" is the only value we need to check.

Reply:
`:type` is for slingshot, we use `:kind` for `ex-info` it looks like (but it's not a requirement). I went ahead and switched it over to ex-info. It was easier to get it working by switching the other error over than trying to figure out how to get slingshot and ex-info to play well together inside the same `try+` (also it's just `try` now 🎉 ). So that change is in a second commit.

PostgreSQL does the normalization `"OFF"` and `"ON"` are lowercased when changed.

I went ahead and left it in this flexible form as it'll make it easier to add settings in the future. I imagine we'll at least want to verify `server_encoding` is `UTF8`.